Add TimescaleDB with AWS Lambda tutorial (#325)
timescale/docs: timescaledb/tutorials/aws-lambda/3rd-party-api-ingest.md

# Pull data from 3rd party API and ingest into TimescaleDB (Docker)
In this section, you will build a data pipeline which pulls data from a 3rd party finance API and loads it into TimescaleDB.

**Required libraries:**

* pandas
* requests
* psycopg2
* pgcopy

If you need to use multiple external libraries - like in this tutorial - your deployment package size can easily be 

> larger than the 250MB limit of Lambda. In this case, you might want to read about. With a Docker container, your 

Author: Right, fixed here https://github.com/timescale/docs/pull/325/commits/d87885bad7bc90e6aea287055f83da469694b81c#

                           'close': 'price_close',
                           'high': 'price_high',
                           'low': 'price_low',
                           'volume': 'trading_volume'}
                           )
   return [row for row in df.itertuples(index=False, name=None)]
 
def handler(event, context):
   symbols = get_symbols()
   for symbol in symbols:
       print("Fetching data for: ", symbol)
       for month in range(1, 2):
           stock_data = fetch_stock_data(symbol, month)
           print('Inserting data...')
           mgr = CopyManager(conn, 'stocks_intraday', columns)
           mgr.copy(stock_data)
           conn.commit()

```

## Add requirements.txt

Add a requirements text file to your project that includes all the libraries that you need installed. For example, 
if you need pandas, requests, psycopg2 and pgcopy this is what your `requirements.txt` should look like:

```
pandas
requests
psycopg2-binary
pgcopy
```

Notice how we use `psycopg2-binary` not `psycopg2`. The binary version of the library contains all its dependencies so 
you don’t need to install them separately.

## Create the Dockerfile

> Dockerfile:

Author: @Loquacity you mean this whole file, or just the Dockerfile step?

--code ImageUri=<ECR Image URI> --role arn:aws:iam::818196790983:role/Lambda

```

## Schedule your Lambda function
If you want to run your Lambda function periodically, you can set up an EventBridge trigger.
Create a new rule with a cron-like expression. For example, if you want to run the function everyday at 9am, 
you can use this expression: `cron(0 9 * * ? *)`.

```bash
aws events put-rule --name schedule-lambda --schedule-expression 'cron(0 9 * * ? *)'
```

If you encounter the `Parameter ScheduleExpression is not valid` error message, have a look at the [cron expression examples in the EventBridge docs](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-create-rule-schedule.html#eb-cron-expressions).

Grant the necessary permissions for the Lambda function.

```bash
aws lambda add-permission --function-name <FUNCTION_NAME> \
--statement-id my-scheduled-event --action 'lambda:InvokeFunction' \
--principal events.amazonaws.com
```

To add the function to the EventBridge rule, create a `targets.json` file containing a memorable, unique string, 
and the ARN of the Lambda Function.
```json
[
  {
    "Id": "docker_lambda_trigger",
    "Arn": "<FUNCTION_ARN>"
  }
]
```

> Then use the `events put-target` command.

Reviewer: ```suggestion
When you have finished, use the `events put-target` command to .... <FIXME>
```

Author: https://github.com/timescale/docs/pull/325/commits/6cdfc2d4613581a6d4e9de396f8f812068116c5b

```bash
aws events put-targets --rule schedule-lambda --targets file://targets.json
```

> To check if the rule is connected to the Lambda function, go to AWS console > Amazon EventBridge > Events > Rules, 

Reviewer: ```suggestion
To check if the rule is connected to the Lambda function, in the AWS console, navigate to Amazon EventBridge > Events > Rules, and
```

Reviewer: Can probably use a proper right arrow here, too, but that's non-blocking.

> select your rule and the Lambda function’s name should be under “Target(s)”.

Reviewer: ```suggestion
select your rule. The Lambda function’s name is listed under `Target(s)`.
```
